apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerCommandTextBuilder.cs

```diff
@@ -255,13 +255,117 @@ public static SpannerCommandTextBuilder FromCommandText(string commandText)
                     }
                     else
                     {
+                        if (removedCommentOrHint)
+                        {
+                            throw new ArgumentException(
+                                $"Spanner {commandName} commands may be specified as '{commandName} <table>' with " +
+                                "parameters added to customize the command with filtering or updated values. " +
+                                "In that case, the commands are sent to Spanner as mutations which don't support comments" +
+                                "or hints, or any other SQL syntax feature. If you need to use comments or hints, " +
+                                "please use DML commands. You can read more about this topic in " +
+                                "https://cloud.google.com/spanner/docs/dml-versus-mutations.");
+                        }
                         targetTable = ValidateTableName(commandSections[1], nameof(commandText));
                         // This just normalizes the command text.
                         commandText = $"{commandName} {targetTable}";
                     }
                     break;
             }
-            return new SpannerCommandTextBuilder(commandText, commandType, targetTable, extraStatements: null);
+            return new SpannerCommandTextBuilder(commandText.Trim(), commandType, targetTable, extraStatements: null);
+        }
+
+        private static string RemoveLeadingCommentsAndHints(string commandText, out bool removedCommentOrHint)
+        {
+            try
+            {
+                return commandText.Substring(IndexOfCommandStart(commandText, out removedCommentOrHint));
+            }
+            catch(Exception ex) when (ex is ArgumentOutOfRangeException || ex is IndexOutOfRangeException)
+            {
+                throw new ArgumentException($"'{commandText}' is not a recognized Spanner command.", nameof(commandText), ex);
+            }
+        }
+
+        private static int IndexOfCommandStart(string command, out bool removedCommentOrHint)
+        {
+            int currentIndex = 0;
+            int loops = 0;
+            do
+            {
+                loops++;
+                SkipWhiteSpaces();
+            }
+            while (SkipLineComment() || SkipBlockComment() || SkipHint());
+
+            removedCommentOrHint = loops > 1;
+            return currentIndex;
+
+            void SkipWhiteSpaces()
+            {
+                while (char.IsWhiteSpace(command[currentIndex]))
+                {
+                    currentIndex++;
+                }
+            }
+
+            bool SkipLineComment()
+            {
+                if (command.IndexOf('#', currentIndex, 1) >= 0 ||
+                    command.IndexOf("--", currentIndex, 2) >= 0)
+                {
+                    // TODO: If this can be escaped, let's check that it's not
```

Comment:
It cannot be escaped, so we don't have to worry about that.

The only situation that we might need to worry about is that there is a statement hint that supports string literals (currently undocumented):

```sql
@{OPTIMIZER_STATISTICS_PACKAGE='auto_20191128_14_47_22UTC'}
SELECT *
FROM AlbumsWithVersion 
where AlbumId = 1
```

The above example is an auto-generated package name which will never contain any characters that will clash with comment or statement hint start or end characters, but if it would also support user-defined package names, we would need to skip string literals in the hints section as well.
(Note: #6848 also currently does not take this fully into account, and would break if a user-defined package name would contain a `}`)

Reply:
@olavloite would you be OK to deferring that check to if and when that is supported? Presumable those package names, even if user defined, might never even accept a }?

Reply:
Yes, I'm fine with that.